c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
 
 class NativeCachingAllocator : public CUDAAllocator {
  private:
-  std::mutex mutex;
+  static constexpr size_t kNumMutexShard = 1031;
+
+  // TODO: change to hardware_destructive_interference_size when C++17 is available.
+  struct alignas(64) AlignedMutex {
+    std::mutex m;
+  };
+
+  std::array<AlignedMutex, kNumMutexShard> mutex;
```

Comment:
Please add a block of comments explaining what we do and why.
Shard allocation region to have independent mutexes to reduce contention.